### ISSUE TITLE
The most efficient ways to state selections?

### ISSUE THREAD

**Issue by martingraham** [arose/ngl#320]
I was setting up a residue spacefill representation with the syntax "(125:A/0 OR 128:A/0 OR 130:A/0 OR ...hundreds more with different chains and models...) AND .CA" and it was taking up to 3.5 seconds for one data set on a large pdb and north of 30 seconds for another on a larger pdb, spending a lot of time in the test routine according to chrome.

A bit of rejigging / guesswork and timing found that restating the selection string as "(/0 AND ( :A AND (125 OR 128 OR 130) AND (:B AND (56  OR 64 OR 86)) .. etc etc for models / chains) AND .CA" - essentially nesting the residue choices within chains and the chains within models - ran a lot faster. Like 200ms instead of 3.5s for my first data set and sub 10 seconds for my second.

I obvs tested the resulting view and all the right residues are in the same places as the old way of doing it.

I guess it depends on how your code is set up (and if it changes in the future), but it might be worth stating on the selection language page that bracketing like this saves a fair bit of time - https://github.com/arose/ngl/blob/master/doc/usage/selection-language.md

(or might already be known by everyone apart from me)

**Comment by arose** [comment]
good point, it would be great to state that "bracketing" usually works faster

can you share your long running example? 30s and also 10s is way too long, hope to be able to to something about it

**Comment by martingraham** [comment]
It looks like whats holding up the really long wait is something else, but I've put the first example in a codepen I forked off someone else's issue in here. I've randomised the residues just so I'm not giving away any trade secrets.

https://codepen.io/anon/pen/bRGbmx

The flat syntax takes ~4.4s seconds to render on my machine, and the nested syntax of the same residues about 0.25 seconds ...

Reversing the order in which they're set up doesn't make a difference

One's in red and the other blue, both 0.5 opacity, which gives all the residues a purple colour. From this I can spot the selections are the same.

**Comment by arose** [comment]
Thanks @martingraham! Is the nested syntax fast enough for you? I guess it is as fast as you can get currently when using the selection language.

**Comment by martingraham** [comment]
Ok, one of the things that was contributing to my lonnnnng wait was setting up / reselections on a label representations.

https://codepen.io/anon/pen/RgPwvM

This example has one label per chain (4v88 has a lot of chains!), indexed at the first atom in a chain. Setting up the representation with the @atom syntax takes about 0.25-0.3s on my machine, the one with the residue:chain syntax takes 2.8-3s... 10 times longer...

This could be because using the residue as the selection means the label routine is trying to calculate labels for each atom in the residue rather than just one (is there a syntax to pick the first atom in a residue?)? Though there is only one label specified per chain in the labeltext object both representations use so both output the same labels.

Anyways, it's a lot faster now :-)

**Comment by martingraham** [comment]
Also, I was trying to build a cache of atom indexes for alpha carbon atoms. I was making a selection for each individual residue i.e. 18:A/0 and .CA and then doing .getAtomIndices() on the selection. Unsurprisingly, making one big selection for all residues got me these indexes a lot quicker... again of the order of 10x quicker

I then used an AtomProxy to get back the residue/chain/model for these atom indices to build the cache (I've discovered some broken pdbs where some residues report not having a .CA atom - these are generally hand-altered pdbs - so i can't rely on the output atom indices being a 1-1 match with the residues in the selection that produced the list). Still a lot faster though.

So basically I find concatenating together one big selection is quicker than a lot of little ones :-)

**Comment by arose** [comment]
So, is your ultimate goal to label each chain in a structure as fast as possible?

**Comment by martingraham** [comment]
Pretty much, one of the requests from a user was to interactively filter out chains from the 3d view that weren't currently of interest, This would involve toggling the presence of the labels for those chains on and off too. The 4v88 pdb is an extreme case, but 3 seconds to recalc some labels with the original selection method was actually taking longer than the hiding/showing of the chains in the structure. If there's a shortcut I don't know of I'll be happy to take it.

**Comment by arose** [comment]
If you just want the first atom of each chain you could loop over the chains and build a `@0,102,2042` atomindex-based selection.

**Comment by arose** [comment]
untested...
```javascript
var idxList = [];
structure.eachChain(function(cp){
  idxList.push( cp.atomOffset );
});
var sele = "@" + idxList.join(",");
```

**Comment by martingraham** [comment]
yeah, that's how I worked out the atom indexes for the codepen example (I console logged them and copied them over to the codepen). If I worked out the halfway point between atomoffset and atomend would the labels appear halfway along each chain (with the same atom indices used in the labeltext parameter too)?

**Comment by arose** [comment]
> If I worked out the halfway point between atomoffset and atomend would the labels appear halfway along each chain (with the same atom indices used in the labeltext parameter too)?

Each residue in a chain may have a different number of atoms but, approximately, yes. However, halfway along a chain can be weird depending on how it folds... makes me want to work on #141 to show labels at the center of residues/chains/models.

**Comment by arose** [comment]
ping @martingraham, good for now?

**Comment by martingraham** [comment]
Yeah, basically the findings for anyone else interested are:

1. concatenating together and running one big selection is quicker than running a lot of little selections separately
2. the nesting model AND ( chain AND ( residue1 OR ... OR residueN )) syntax is quicker than the OR'ing lots of residue:chain/model syntax
3. one residue range selection is quicker than OR'ing lots of consecutive residues in the same selection
4. selections with atom '@' indexes are very quick
